### PR TITLE
Feature/login

### PR DESCRIPTION
**Pull request by kobyfogel**

### CODE REVIEW

File: app/internal/security/dependancies.py [hunk 1]
```diff
@@ -23,3 +25,23 @@ async def is_manager(
     """
     await check_jwt_token(db, jwt, manager=True)
     return True
+
+
+async def current_user(
+    request: Request,
+        db: Session = Depends(get_db),
+        jwt: str = Depends(get_authorization_cookie),
```
Comment (yammesicka) [inline app/internal/security/dependancies.py:33]: indentation

File: app/internal/security/dependancies.py [hunk 2]
```diff
@@ -1,8 +1,10 @@
 from starlette.requests import Request
 
+from app.database.models import User
 from app.dependencies import get_db
 from app.internal.security.ouath2 import (
-    Depends, Session, check_jwt_token, get_authorization_cookie)
+    Depends, Session, check_jwt_token, get_authorization_cookie,
+    HTTPException, HTTP_401_UNAUTHORIZED)
```
Comment (yammesicka) [inline app/internal/security/dependancies.py:7]: Close in a separate line, order the imports

File: app/static/style.css [hunk 3]
```diff
@@ -96,6 +96,16 @@ p {
...
 }
+  
+.error-message {
+    line-height: 0;
+    color: red;
+    padding-left: 12.5rem;
+}
+
+.subtitle {
+    font-size: 20px;
```
Comment (yammesicka) [inline app/static/style.css:107]: Use rem

File: app/internal/security/ouath2.py [hunk 4]
```diff
@@ -70,9 +70,9 @@ async def check_jwt_token(
         jwt_payload = jwt.decode(
             token, JWT_KEY, algorithms=JWT_ALGORITHM)
         if not manager:
-            return True
+            return jwt_payload.get("sub"), jwt_payload.get("user_id")
         if jwt_payload.get("is_manager"):
-            return True
+            return jwt_payload.get("sub"), jwt_payload.get("user_id")
```
Comment (yammesicka) [inline app/internal/security/ouath2.py:75]: Use `or`

File: tests/security_testing_routes.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
-from app.internal.security.dependancies import (is_manager, is_logged_in)
+from app.internal.security.dependancies import (
+    current_user, is_manager, is_logged_in, User)
```
Comment (yammesicka) [inline tests/security_testing_routes.py:2]: Close in a separate line

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/security/dependancies.py (hunk 2 above)
Comment (ivarshav) [inline app/internal/security/dependancies.py:7]: It's a bit weird to import HTTP_401_UNAUTHORIZED from this path

File: tests/security_testing_routes.py [hunk 6]
```diff
@@ -31,3 +32,12 @@ async def manager_route(
     # if user not looged-in, or have no manager permission,
     # will be redirected to login route.
     return {"manager": user}
+
+
+@router.get('/current_user')
+async def logged_in_user(
+        request: Request, user: User = Depends(current_user)):
+    # This is how to protect route for logged in user only.
```
Comment (ivarshav) [inline tests/security_testing_routes.py:40]: use """ instead of #

**Review by PureDreamer — COMMENTED** [review]
I think you need to add the navbar to present only the sign-out if the user is logged in and to show only the sign in, sign up if he is not connected.

File: app/static/style.css (hunk 3 above)
Comment (kobyfogel) [reply app/static/style.css:107]: that's not my code.. must be from merging..

File: app/internal/security/dependancies.py [hunk 7]
```diff
@@ -1,8 +1,12 @@
+from fastapi import HTTPException
+from starlette.status import HTTP_401_UNAUTHORIZED
 from starlette.requests import Request
 
+from app.database.models import User
 from app.dependencies import get_db
 from app.internal.security.ouath2 import (
-    Depends, Session, check_jwt_token, get_authorization_cookie)
+    Depends, check_jwt_token, get_authorization_cookie, Session
```
Comment (yammesicka) [inline app/internal/security/dependancies.py:8]: Order by abc

File: app/internal/security/reset_password.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    html = template.render(
+         recipient=user.username,
+         site_name=CALENDAR_SITE_NAME,
+         link=f"{CALENDAR_RESET_PASSWORD_PAGE}?token={user.token}",
+         email=user.email)
+    message = MessageSchema(
+        subject=f"{CALENDAR_SITE_NAME} reset password",
+        recipients=[user.email], body=html, subtype='html')
+    background_tasks.add_task(mail.send_message, message)
+    return True
```
Comment (yammesicka) [inline app/internal/security/reset_password.py:27]: Please use the already existing `send_mail`

File: app/internal/security/ouath2.py [hunk 9]
```diff
@@ -69,10 +69,8 @@ async def check_jwt_token(
     try:
         jwt_payload = jwt.decode(
             token, JWT_KEY, algorithms=JWT_ALGORITHM)
-        if not manager:
-            return True
-        if jwt_payload.get("is_manager"):
-            return True
+        if not manager or jwt_payload.get("is_manager"):
```
Comment (yammesicka) [inline app/internal/security/ouath2.py:72]: This line is problematic. Might be a simple naming issues - but currently it contains a logic contradiction.

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/security/dependancies.py (hunk 2 above)
Comment (efratush) [reply app/internal/security/dependancies.py]: Why not import HTTP directly?

File: app/internal/security/dependancies.py [hunk 10]
```diff
@@ -1,8 +1,12 @@
+from fastapi import HTTPException
+from starlette.status import HTTP_401_UNAUTHORIZED
 from starlette.requests import Request
 
+from app.database.models import User
 from app.dependencies import get_db
 from app.internal.security.ouath2 import (
-    Depends, Session, check_jwt_token, get_authorization_cookie)
+    Depends, Session, get_jwt_token, get_authorization_cookie
```
Comment (efratush) [inline app/internal/security/dependancies.py:8]: You will also transfer Depends for import from fastapi.

File: app/internal/security/dependancies.py [hunk 11]
```diff
@@ -21,5 +25,32 @@ async def is_manager(
...
+    """
+    Returns logged in User object.
+    A dependency function protecting routes for only logged in user.
+    """
+    jwt_payload = await get_jwt_token(db, jwt)
+    username = jwt_payload.get("sub")
+    user_id = jwt_payload.get("user_id")
+    db_user = await User.get_by_username(db, username=username)
+    if db_user and db_user.id == user_id:
+        return db_user
```
Comment (yammesicka) [inline app/internal/security/dependancies.py:51]: Prefer to returning a dictionary with the user's details here.
We don't want to inquire the DB every time someone wants to use the current_user dependency :)

You can also create another function, `current_user_from_db` to supply more options

**Review by yammesicka — CHANGES_REQUESTED** [review]
Great progress. Few additional insights

File: app/internal/security/dependancies.py [hunk 12]
```diff
@@ -21,5 +25,32 @@ async def is_manager(
...
+    jwt_payload = await get_jwt_token(db, jwt)
+    if jwt_payload.get("is_manager"):
+        return True
+    raise HTTPException(
+                status_code=HTTP_401_UNAUTHORIZED,
+                headers=request.url.path,
+                detail="You don't have a permition to enter this page")
+
+
+async def current_user(
```
Comment (Gonzom) [inline app/internal/security/dependancies.py:37]: "current_user" -> "get_current_user"

Comment (kobyfogel) [reply app/internal/security/dependancies.py]: That is a dependency function, and not a function that anyone supposed to call directly. After many discussions, this is the name that was decided for this dependency.

**Review by yammesicka — APPROVED** [review]

**Pull request merged**